sycl/test/hier_par/hier_par_basic.cpp

```diff
@@ -0,0 +1,183 @@
+//==- hier_par_basic.cpp --- hierarchical parallelism API test -------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// RUN: %clang -std=c++11 -fsycl %s -o %t.out -lOpenCL -lstdc++
+// RUN: %t.out
+
+// This test checks hierarchical parallelism invocation APIs, but without any
+// data or code with side-effects between the work group and work item scopes.
+
+#include <CL/sycl.hpp>
+#include <iostream>
+#include <memory>
+
+using namespace cl::sycl;
+
+template <typename GoldFnTy>
+bool verify(int testcase, int range_length, int *ptr, GoldFnTy get_gold) {
+  int err_cnt = 0;
+
+  for (int i = 0; i < range_length; i++) {
+    int gold = get_gold(i);
+
+    if (ptr[i] != gold) {
+      if (++err_cnt < 20) {
+        std::cout << testcase << " - ERROR at " << i << ": " << ptr[i]
+                  << " != " << gold << "(expected)\n";
+      }
+    }
+  }
+  if (err_cnt > 0)
+    std::cout << "-- Failure rate: " << err_cnt << "/" << range_length << "("
+              << err_cnt / (float)range_length * 100.f << "%)\n";
+  return err_cnt == 0;
+}
+
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+
+int main() {
+  const int N_WG = 7;
+  const int WG_SIZE_MAIN = 3;
+  const int WG_SIZE1 = 5;
+  const int WG_SIZE2 = 2;
+  const int N_ITER = 1000;
+
+  const int range_length = N_WG * WG_SIZE_MAIN;
+  std::unique_ptr<int> data(new int[range_length]);
+  int *ptr = data.get();
+  bool passed = true;
+
+  try {
+    queue myQueue;
+
+    std::cout << "Running on "
+              << myQueue.get_device().get_info<cl::sycl::info::device::name>()
+              << "\n";
+    {
+      // Testcase1
+      // - handler::parallel_for_work_group w/o local size specification +
+      //   group::parallel_for_work_item w/o flexible range
+      // - h_item::get_global_id
+      // The global size is not known, so kernel code needs to adapt
+      // dynamically, hence the complex loop bound and index computation.
+      std::memset(ptr, 0, range_length * sizeof(ptr[0]));
```

Comment:
At least it is useful to have some tests using C-style and pointers to be sure SYCL can work with C programs... :-)

Reply:
Right. Let's say the test aims to be closer to what usual SYCL developer would write or port :-)

Reply:
You are depressing me ;-)